python/ray/serve/metric/exporter.py

```diff
@@ -0,0 +1,144 @@
+from typing import Dict
+from collections import Counter, namedtuple
+
+import ray
+from ray.serve.metric.types import MetricType, MetricMetadata, MetricBatch
+from ray.serve.utils import _get_logger
+
+logger = _get_logger()
+
+
+def make_metric_namedtuple(metric_metadata: MetricMetadata,
+                           record: MetricBatch):
+    fields = ["name", "type"]
+    fields += list(metric_metadata.default_labels.keys())
+    fields += list(record.labels.keys())
+
+    merged_labels = metric_metadata.default_labels.copy()
+    merged_labels.update(record.labels)
+
+    tuple_type = namedtuple(metric_metadata.name, fields)
+    return tuple_type(
+        name=metric_metadata.name, type=metric_metadata.type, **merged_labels)
+
+
+class BaseExporter:
+    """Aggregate metrics from all RayServe actors and export them.
+
+    Metrics are aggregated pushed from other actors in the system to
+    this actor. They can then be stored or pushed to an external monitoring
+    system.
+    """
+
+    def __init__(self):
+        # Stores the mapping metric_name -> MetricMetadata
+        # This field is tolerant to failures since each client will always push
+        # an updated copy of the metadata.
+        self.metric_metadata = dict()
+
+        logger.debug("Initialized with metric exporter of type {}".format(
+            type(self)))
+
+    def ingest(self, metric_metadata: Dict[str, MetricMetadata],
+               batch: MetricBatch):
+        self.metric_metadata.update(metric_metadata)
+        self.export(self.metric_metadata, batch)
+
+    def export(self, metric_metadata, metric_batch):
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+    def inspect_metrics(self):
+        raise NotImplementedError(
+            "This method should be implemented by subclass.")
+
+
+@ray.remote(num_cpus=0)
+class InMemoryExporter(BaseExporter):
```

Comment:
Can we use composition instead of inheritance here? I don't think inheritance buys us anything and is a worse user interface IMO